### PR TITLE
Call-graph based Unity.Burst support

### PR DESCRIPTION
**Pull request by Domonion**
Support for Unity.Burst

highlights access to managed objects, object creation etc
transitively propagates burst contexts
highlighting are taken and annotated to match Burst compiler

### CODE REVIEW

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs [hunk 1]
```diff
@@ -147,27 +163,75 @@ public override void Execute(Action<DaemonStageResult> committer)
...
 
-        private List<UnityProblemAnalyzerContext> GetProblemAnalyzerContext(ITreeNode element)
+        private UnityProblemAnalyzerContext GetProblemAnalyzerContext(ITreeNode element)
         {
-            var res = new List<UnityProblemAnalyzerContext>();
+            var res = new UnityProblemAnalyzerContext();
             if (myIsPerformanceAnalysisEnabled && IsPerformanceCriticalDeclaration(element))
-                res.Add(UnityProblemAnalyzerContext.PERFOMANCE_CONTEXT);
+                res |= UnityProblemAnalyzerContext.PERFOMANCE_CONTEXT;
+            if (myIsBurstAnalysisEnabled && IsBurstDeclaration(element))
```
[the remark this answers is not in the document]
Comment (krasnotsvetov) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs:180]: PERFOMANCE -> PERFORMANCE

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                    var visited = new HashSet<ITypeElement>();
+                    var interfaces = new LocalList<ITypeElement>();
+                    var todo = new Stack<ITypeElement>();
+                    visited.Add(@struct);
+                    todo.Push(@struct);
+
+                    while (!todo.IsEmpty())
+                    {
+                        var current = todo.Pop();
+                        foreach (var typeElement in current.GetSuperTypeElements())
```
Comment (krasnotsvetov) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs:51]: What's about using GetAllSuperTypes and remove while cyle?

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            return typeElement is IClass @class && (@class.IsSystemValueTypeClass() || @class.IsObjectClass());
+        }
+
+        [ContractAnnotation("null => false")]
+        public static bool IsBurstProhibitedMethod(this IFunction function)
+        {
+            if (function == null)
+                return false;
+            var containingType = function.GetContainingType();
+            if (containingType is IStruct && function is IMethod method && method.IsOverridesObjectGetHashCode())
```
Comment (krasnotsvetov) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs:40]: What's about overriding equals?

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+                    conditionalAccessExpression.ConditionalQualifier as IConditionalAccessExpression;
+            }
+
+            return conditionalAccessExpression?.ConditionalQualifier?.Type().IsOpenType ?? false;
+        }
+    
+        private static bool MayContainProhibitedMethods(this ITypeElement typeElement)
+        {
+            //                                             CGTD is valueType is what i want? what is this anyway?
+            return typeElement is IClass @class && (@class.IsSystemValueTypeClass() || @class.IsObjectClass());
```
Comment (krasnotsvetov) [inline resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs:31]: Looks like magic, could you add comments?

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstStringLiteralOwnerAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+{
+    [SolutionComponent]
+    public class BurstStringLiteralOwnerAnalyzer : BurstProblemAnalyzerBase<ICSharpLiteralExpression>
+    {
+        protected override void Analyze(ICSharpLiteralExpression cSharpLiteralExpression, IDaemonProcess daemonProcess, DaemonProcessKind kind,
+            IHighlightingConsumer consumer)
+        {
+            if (cSharpLiteralExpression.Literal.GetTokenType().IsStringLiteral)
+            {
+                consumer.AddHighlighting(new BC1033Error(cSharpLiteralExpression.GetDocumentRange()));
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstStringLiteralOwnerAnalyzer.cs:18]: Known issue, WIP

File: resharper/resharper-unity/test/data/CSharp/Daemon/Stages/BurstCodeAnalysis/resharper/ComplexTest.cs [hunk 6]
```diff
@@ -0,0 +1,271 @@
+using System;
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/test/data/CSharp/Daemon/Stages/BurstCodeAnalysis/resharper/ComplexTest.cs:1]: WIP

[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/test/data/CSharp/Daemon/Stages/BurstCodeAnalysis/resharper/ComplexTest.cs:1]: fixed

File: resharper/resharper-unity/test/data/CSharp/Daemon/Stages/BurstCodeAnalysis/resharper/ComplexTest.cs [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using Unity.Burst;
+using Unity.Collections;
+using Unity.Jobs;
+// ReSharper disable UnusedMember.Global
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/test/data/CSharp/Daemon/Stages/BurstCodeAnalysis/resharper/ComplexTest.cs]: fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstForeachStatementAnalyzer.cs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    [SolutionComponent]
+    public class BurstForeachStatementAnalyzer : BurstProblemAnalyzerBase<IForeachStatement>
+    {
+        protected override void Analyze(IForeachStatement foreachStatement, IDaemonProcess daemonProcess,
+            DaemonProcessKind kind, IHighlightingConsumer consumer)
+        {
+            //juding by latest documentation, foreach is prohibited to use.
+            //burst compiler does not allow to use it either, even with struct enumerator
+            //despite all of that, foreach is used in unity sources with burst ¯\_(ツ)_/¯
+            consumer.AddHighlighting(new BC1037Error(foreachStatement.ForeachKeyword.GetDocumentRange()));
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstForeachStatementAnalyzer.cs:19]: fixed

File: resharper/resharper-unity/src/Settings/UnitySettings.cs [hunk 9]
```diff
@@ -37,6 +37,9 @@ public class UnitySettings
         // Analysis
         [SettingsEntry(true, "Enables performance analysis in frequently called code")]
         public bool EnablePerformanceCriticalCodeHighlighting;
+        
+        [SettingsEntry(true, "Enables analysis in Burst code")]
+        public bool EnableBurstCodeHighlighting;
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/Settings/UnitySettings.cs:51]: Added

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs]: renamed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs [hunk 10]
```diff
@@ -186,11 +250,22 @@ public override void ProcessBeforeInterior(ITreeNode element, IHighlightingConsu
...
+                    if (myProhibitedContexts.Count > 0)
+                        prohibitedContexts = myProhibitedContexts.Peek();
+                    for (byte context = 1, index = 0; index < end; index++, context <<= 1)
                     {
-                        foreach (var performanceProblemAnalyzer in myProblemAnalyzersByContext[context])
+                        var enumContext = (UnityProblemAnalyzerContext) context;
+                        if (possibleContexts.HasFlag(enumContext) && !prohibitedContexts.HasFlag(enumContext))
                         {
-                            performanceProblemAnalyzer.RunInspection(element, DaemonProcess, myProcessKind, consumer);
+                            //be aware of https://johnthiriet.com/back-to-basics-csharp-casting-an-integer-into-an-enumeration/
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs]: removed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        }
+
+        public override LocalList<IDeclaredElement> GetRootMarksFromNode(ITreeNode currentNode,
+            IDeclaredElement containingFunction)
+        {
+            var result = new HashSet<IDeclaredElement>();
+            switch (currentNode)
+            {
+                #if JET_MODE_ASSERT
+                case IMethodDeclaration methodDeclaration when methodDeclaration.DeclaredElement is IMethod method &&
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs]: removed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                }
+                #endif
+                case IStructDeclaration structDeclaration
+                    when structDeclaration.DeclaredElement is IStruct @struct &&
+                         @struct.HasAttributeInstance(KnownTypes.BurstCompileAttribute, AttributesSource.Self):
+                {
+                    var visited = new HashSet<ITypeElement>();
+                    var interfaces = new LocalList<ITypeElement>();
+                    var todo = new Stack<ITypeElement>();
+                    visited.Add(@struct);
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs]: redesigned, removed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/CallGraph/ExpensiveCodeCallGraphAnalyzer.cs [hunk 13]
```diff
@@ -22,20 +23,62 @@ public class ExpensiveCodeCallGraphAnalyzer : CallGraphRootMarksProviderBase
...
+            if (containingFunction == null)
+                return result;
+
+            var declaration = currentNode as IDeclaration;
+            var declaredElement = declaration?.DeclaredElement;
+            if (!ReferenceEquals(containingFunction, declaredElement))
+                return result;
+            
+            var processor = new ExpensiveCodeProcessor();
+            declaration.ProcessThisAndDescendants(processor);
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/CallGraph/ExpensiveCodeCallGraphAnalyzer.cs]: redesigned, fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/CallGraph/PerformanceCriticalCodeCallGraphMarksProvider.cs [hunk 14]
```diff
@@ -105,5 +91,55 @@ private IDeclaredElement ExtractMethodDeclarationFromStartCoroutine([NotNull]ICS
...
+                result.Add(coroutineOrInvoke);
+
+            if (containingFunction == null)
+                return result;
+            var declaration = currentNode as IDeclaration;
+            var declaredElement = declaration?.DeclaredElement;
+            if (!ReferenceEquals(containingFunction, declaredElement))
+                return result;
+            var processor = new PerformanceCriticalCodeProcessor();
+            declaration.ProcessThisAndDescendants(processor);
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/PerformanceCriticalCodeAnalysis/CallGraph/PerformanceCriticalCodeCallGraphMarksProvider.cs:111]: redesigned fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs (hunk 2 above)
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs]: switched, fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs (hunk 4 above)
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs]: redesigned, comments added

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs (hunk 3 above)
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/BurstCodeAnalysisUtil.cs:47]: checked, tested, fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstStringLiteralOwnerAnalyzer.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/Analyzers/BurstStringLiteralOwnerAnalyzer.cs:18]: checked, tested, fixed

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs [hunk 15]
```diff
@@ -36,12 +39,14 @@ public abstract class UnityHighlightingAbstractStage : CSharpDaemonStageBase
 
         protected UnityHighlightingAbstractStage(CallGraphSwaExtensionProvider callGraphSwaExtensionProvider,
             PerformanceCriticalCodeCallGraphMarksProvider performanceCriticalCodeCallGraphMarksProvider,
+            CallGraphBurstMarksProvider callGraphBurstMarksProvider,
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/UnityHighlightingAbstractStage.cs:42]: postponed till another analysis.

File: resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+    </Message>
+    <Range>Range</Range>
+    <Behavour overlapResolvePolicy="NONE" />
+  </Error>
+
+  <Error name="BC1042Short" configurableSeverity="Unity.BC1042Short">
+    <Parameter type="DocumentRange" name="range" />
+    <Parameter type="string" name="typeQualification"/>
+    <Message value="BC1042: The managed class type `{0}` is not supported.">
+      <Argument>TypeQualification</Argument>x
```
Comment (citizenmatt) [inline resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml:194]: Extra `x` 😄

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+using static JetBrains.ReSharper.Daemon.CSharp.CallGraph.CallGraphCSharpUtil;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.BurstCodeAnalysis.CallGraph
+{
+    [SolutionComponent]
+    public class CallGraphBurstMarksProvider : CallGraphRootMarksProviderBase
+    {
+        public CallGraphBurstMarksProvider(ISolution solution)
+            : base(nameof(CallGraphBurstMarksProvider),
+                new CallGraphOutcomingPropagator(solution, nameof(CallGraphBurstMarksProvider)))
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/BurstCodeAnalysis/CallGraph/CallGraphBurstMarksProvider.cs]: `Outcoming` was chosen to show that it is opposite to`IncomingPropagator`, so idk if this is correct change

File: resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs [hunk 18]
```diff
@@ -23,12 +23,16 @@ public class ReSharperOptionsPage : OptionsPageBase
...
+        private static readonly Expression<Func<UnitySettings, bool>> ourEnableBurstHighlightingAccessor =
+            s => s.EnableBurstCodeHighlighting;
 
         public ReSharperOptionsPage(Lifetime lifetime, [NotNull] OptionsSettingsSmartContext settingsStore,
                                     RunsProducts.ProductConfigurations productConfigurations)
             : base(lifetime, settingsStore)
         {
             Header("C#");
+            CheckBox((UnitySettings s) => s.EnableBurstCodeHighlighting,
+                "Enables Burst code analysis");
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/Application/UI/Options/VisualStudio/ReSharperOptionsPage.cs:35]: fixed, thx

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingGroupIds.cs [hunk 19]
```diff
@@ -6,9 +6,11 @@ namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Highlightings
     // RegisterConfigurableHighlightingsGroup registers a group in Inspection Severity
     [RegisterConfigurableHighlightingsGroup(Unity, "Unity")]
     [RegisterConfigurableHighlightingsGroup(UnityPerformance, "Unity Performance Inspections", PencilsGroupKind.UnityPerformanceKind)]
+    [RegisterConfigurableHighlightingsGroup(Burst, "Burst")]
     public static class UnityHighlightingGroupIds
     {
         public const string Unity = "UNITY";
         public const string UnityPerformance = "UNITY_PERFORMANCE";
+        public const string Burst = "Burst";
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingGroupIds.cs:14]: fixed, thx

File: resharper/resharper-unity/src/Rider/UnityOptionsPage.cs [hunk 20]
```diff
@@ -44,6 +47,8 @@ public class UnityOptionsPage : OptionsPageBase
             CheckBox((UnitySettings s) => s.AllowAutomaticRefreshInUnity, "Automatically refresh assets in Unity");
 
             Header("C#");
+            CheckBox(ourEnableBurstHighlightingAccessor,
+                "Enables Burst code analysis");
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/Rider/UnityOptionsPage.cs:51]: fixed, thx

File: resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml (hunk 16 above)
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml]: Fixed, thx

File: resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml [hunk 21]
```diff
@@ -0,0 +1,199 @@
...
+    JetBrains.ReSharper.Psi.CSharp.Tree;
+    JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+    JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Highlightings;
+    JetBrains.ReSharper.Plugins.Unity.CSharp.Psi.Resolve;
+    JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.PerformanceCriticalCodeAnalysis.Highlightings;
+  </Usings>
+
+  <SeverityConfiguration>
+    <Group name="UnityHighlightingGroupIds.Burst">
+      <Tag externalName="BC1001Error.HIGHLIGHTING_ID" default="WARNING">
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Errors/BurstErrors.xml:16]: Burst compiler is a special case: if compiled at the editor - despite unsuccessful compilation unity project will be executed, if built as standalone - unity will consider it as a normal error, so I think warning is the best severity.

File: resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingGroupIds.cs [hunk 22]
```diff
@@ -6,9 +6,11 @@ namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Highlightings
     // RegisterConfigurableHighlightingsGroup registers a group in Inspection Severity
     [RegisterConfigurableHighlightingsGroup(Unity, "Unity")]
     [RegisterConfigurableHighlightingsGroup(UnityPerformance, "Unity Performance Inspections", PencilsGroupKind.UnityPerformanceKind)]
+    [RegisterConfigurableHighlightingsGroup(Burst, "Burst")]
```
[the remark this answers is not in the document]
Comment (Domonion) [reply resharper/resharper-unity/src/CSharp/Daemon/Stages/Highlightings/UnityHighlightingGroupIds.cs]: Renamed, yes they should, explained higher.

**Pull request merged**